Add/updated creds flow calypso

vianasw opened the pull request:
#### Changes proposed in this Pull Request

* This patch implements stream processing of `application/x-ndjson responses`. It is a port of the code we used in the `wpcom-xhr-request` library (see [PR here](https://github.com/Automattic/wpcom-xhr-request/pull/31)). We've also brought the advanced-credentials component from jetpack-cloud.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* To test this PR you'll need to apply this diff in your wpcom sandbox: D75408-code.

* Then, change your `/etc/hosts` file to point to your wpcom sandbox.

* Start calypso locally: `yarn start`

* Create a JN site and add a Backup plan to it.

* Then go to http://calypso.localhost:3000/settings/jetpack/<JN_site_url>#credentials

* And test the credentials flow with valid and invalid credentials. It should behave the same way as in jetpack-cloud.

Commit:
Move creds component from calypso green

Migrate credentials form component used in Calypso green to blue.

vianasw committed:
Process ndjson requests in stream mode

This is a work in progress to process ndjson requests.

vianasw committed:
Fix infinite loop when processing ndjson messages

We're only sending the response chunks to onStreamRecord when we get
the 207 status code.

vianasw committed:
Code style fix

vianasw committed:
Bring component changes from jetpack-cloud

Bringing changes after a refactor in the host-selection component in
jetpack-cloud.

jsnajdr commented on client/components/advanced-credentials/credentials-form/style.scss:
```diff
@@ -0,0 +1,125 @@
+		}
+	}
+
+	&__buttons .credentials-form__buttons-flex {
```
Let's write this as expanded `.credentials-form__buttons`. The `&__buttons` shortcut makes it very difficult to grep for usages of the CSS class name.

jsnajdr commented on client/components/advanced-credentials/connection-status.tsx:
```diff
@@ -0,0 +1,59 @@
+	state: StatusState;
+}
+
+const ConnectionStatus: FunctionComponent< Props > = ( { state } ) => {
```
Are you duplicating all the `advanced-credentials` Jetpack Cloud components here? There should be no need for the duplication, just move the files to the common directory and update existing imports.

jsnajdr commented on client/my-sites/site-settings/settings-jetpack/controller.js:
```diff
@@ -2,6 +2,6 @@ import { createElement } from 'react';
 
 export function jetpack( context, next ) {
-	context.primary = createElement( JetpackMain );
+	context.primary = createElement( JetpackMain, { context } );
```
Instead of passing the entire `context` object as a prop, can you pass just the `query`?

Also, any component can read the current route information with the `getCurrentQueryArguments` Redux selector.

Commit:
Use getCurrentQueryArguments and expanded CSS classes

- Use getCurrentQueryArguments instead of passing context.
- Expand CSS classes to simplify searching.

Commit:
Use getCurrentQueryArguments instead of passing context

nunyvega commented:
Thanks for your review @jsnajdr !
I modified the code to use `getCurrentQueryArguments` and expanded CSS instead of using `&`. 
We'll now check if it's possible to use the same component on JP Cloud and here.

Commit:
Use Advanced-credentials common component

and delete duplicate component from Calypso green.

nunyvega commented:
We modified Calypso Green to use the shared component and removed the duplicate code.

jsnajdr commented on client/components/advanced-credentials/host-info.ts:
```diff
@@ -16,23 +15,23 @@ enum InfoTypes {
 export interface Text {
 	type: InfoTypes.Text;
-	text: ReactChild;
+	text: string;
```
You don't want to do any changes in this file. The `ReactChild` type is correct, as the text can be a React markup (with HTML), not just a string. And there shouldn't be any `translate( ... ).toString()`, just `translate()`.

jsnajdr commented on client/components/advanced-credentials/credentials-form/index.tsx:
```diff
@@ -50,7 +50,7 @@ const ServerCredentialsForm: FunctionComponent< Props > = ( {
 					...formState,
 					protocol: currentTarget.value as 'ftp' | 'ssh',
 					port: ( () => {
-						let port = parseInt( String( formState.port ) );
```
What's the justification for removing `String()` here?

jsnajdr commented on client/components/advanced-credentials/verification/index.tsx:
```diff
@@ -18,8 +18,6 @@ interface Props {
 }
 
-const ERROR_CODES_FOR_BLOCKED_REQUEST = [ 'service_unavailable', 'invalid_credentials' ];
-
```
Is this change relevant to this PR? It removes special error reporting for these two errors.

jsnajdr commented on client/my-sites/site-settings/settings-jetpack/main.jsx:
```diff
@@ -1,23 +1,27 @@
 import Main from 'calypso/components/main';
 import SidebarNavigation from 'calypso/my-sites/sidebar-navigation';
-import JetpackCredentials from 'calypso/my-sites/site-settings/jetpack-credentials';
+//import JetpackCredentials from 'calypso/my-sites/site-settings/jetpack-credentials';
```
This import should be removed rather than commented out.

jsnajdr commented on client/my-sites/site-settings/settings-jetpack/main.jsx:
```diff
@@ -1,23 +1,27 @@
 import { getSelectedSite, getSelectedSiteId } from 'calypso/state/ui/selectors';
 
 const SiteSettingsJetpack = ( { site, siteId, siteIsJetpack, showCredentials, translate } ) => {
+	const { host, action } = useSelector( getCurrentQueryArguments );
```
Let's not mix `connect` and `useSelector` in one component. Either add a new selector call to the `connect` map function, or convert the entire component to hooks.

Commit:
Remove new code overwrite and correct typos

- Remove the overwrites of newer code incorrectly
added while migrating component with outdated code.
- Correct typos.

nunyvega commented:
I was mistakenly overwriting the latest changes on this code, which was removing `ERROR_CODES_FOR_BLOCKED_REQUEST` and removing `String` from `let port`, basically reverting the code to the previous version of it. Thanks for catching that, and for the thorough review @jsnajdr !
I corrected that and added the query arguments to the `connect` map function.

jsnajdr commented on packages/wpcom-proxy-request/src/index.js:
```diff
@@ -18,6 +18,8 @@ const proxyOrigin = 'https://public-api.wordpress.com';
  */
 const origin = window.location.protocol + '//' + window.location.host;
 
+let onStreamRecord = null;
```
This is a global variable, but it would be shared by all requests running in parallel. It's better to store the callback as `xhr.onStreamRecord = ...`, then store the `xhr` in `requests[ id ]`, and then retrieve it before called:
```js
if ( statusCode === 207 ) {
  const onStreamRecord = xhr.onStreamRecord;
  if ( onStreamRecord ) {
    onStreamRecord( body );
  }
}
```

jsnajdr reviewed (commented):
There's one bug in `wpcom-proxy-request`, the rest of PR is OK :+1: Thanks for addressing all feedback!

vianasw committed:
Fix tracks event name

The event jetpack_advanced_credentials_flow_host_select was ignored
because it did not match /^calypso(?:_[a-z0-9]+){2,}$/ (not starting
with the calypso prefix).
